core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala

```diff
@@ -183,26 +184,22 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
                                      numRecords: Int,
                                      maxPollRecords: Int = Int.MaxValue): ArrayBuffer[ConsumerRecord[K, V]] = {
     val records = new ArrayBuffer[ConsumerRecord[K, V]]
-    val maxIters = numRecords * 300
-    var iters = 0
-    while (records.size < numRecords) {
-      val polledRecords = consumer.poll(50).asScala
-      assertTrue(polledRecords.size <= maxPollRecords)
-      for (record <- polledRecords)
-        records += record
-      if (iters > maxIters)
-        throw new IllegalStateException("Failed to consume the expected records after " + iters + " iterations.")
-      iters += 1
+    def pollCondition(polledRecords: ConsumerRecords[K, V]): Boolean = {
```

Comment:
Nit: it's confusing to call this functions `condition` given that they often involve side-effects. It's an existing issue since that's the parameter name of `waitUntilTrue`. Maybe `action` would be better?